Show an Open Wallet dialog upon launching the app

wukong1971 opened the pull request:
Updated PR Description
---

### The Open Wallet dialog

1. Upon launching the app, an Open Wallet dialog is displayed, allowing the user to enter the passphrase directly and press `Enter` key or the "OK" button to open the default wallet.

![image](https://user-images.githubusercontent.com/87334822/126019254-2b96c1ce-08f9-4c56-af4d-63c9dd5549a3.png)

2. User can use the "Choose..." button to select a different wallet to open.
3. User can press `ESC` key or the "Cancel" button to dismiss the Dialog.
4. The OpenWallet dialog is a Modal dialog, user has to dismiss the dialog first before they can interact with the main window.

### Next step improvements

1. Show all the error messages, like "Wrong Password", "File not exist", "Invalid wallet file" etc., within the same dialog, and allow the user to make changes to their decisions in the same dialog when these errors occur.
1. Update the UX of menu action "Wallet -> Load" to also use this new Open Wallet dialog.
1. Add a "Create Wallet" button to the dialog so that user can choose to create a new wallet as well in the same dialog. However, before that, **we need a redesign of the existing UX about creating a new wallet first.** The current UX consists of a series of different popups, ideally, we would show all the steps in the same dialog so that user doesn't have to jump from popup to popup.
1. The OpenWallet dialog is pre-populated with `wallet.jmdat`. We should remember the last-opened wallet and show that instead.
1. Consider delay the showing of the main window until after the user has made a decision with the "Open Wallet Dialog"



Original Draft PR Description Below 
--- 

I'm trying to implement a "open/create wallet" dialog (#930) to improve the user experience. And I think it would be a good idea for us to utilize Qt Designer for the UI Design in general:

![image](https://user-images.githubusercontent.com/87334822/125917665-41ac61fb-0a6e-49c9-ba96-3c2ca07983c3.png)

The "open wallet dialog" looks like this when running:
![image](https://user-images.githubusercontent.com/87334822/125919491-12da9f6a-dcf0-4974-a6e6-4e11e5d8ccff.png)


I believe this will make future adjustments of the UI easier, and it gives us a nice separation between the program logic and the UI layout.

However, before moving forward, I'd like to make sure I'm on the right track and that we can agree on the utilization of Qt Designer. So please let me know if you have any concerns.

(The `open_wallet_dialog.py` is generated by `pyside2-uic` based on `open_wallet_dialog.ui`, and that `open_wallet_dialog.ui` can be opened directly in Qt Designer, which can be installed on all major desktop platforms: Windows/macOS/Linux)  

Thanks!

wukong1971 commented on jmui/jmui/open_wallet_dialog.py:
```diff
@@ -0,0 +1,126 @@
+# -*- coding: utf-8 -*-
```
This file is generated based on `open_wallet_dialog.ui`, but the python format is much better for the code review process than the original `.ui` file.

wukong1971 commented on jmui/jmui/open_wallet_dialog.ui:
```diff
@@ -0,0 +1,185 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This file can be opened via "Qt Designer", which is available on all desktop platforms. On Ubuntu, it can be installed by `sudo apt install qttools5-dev-tools`, and opened by running ` /usr/lib/x86_64-linux-gnu/qt5/bin/designer` after the installation.

wukong1971 commented on jmui/setup.py:
```diff
@@ -0,0 +1,16 @@
+      author_email='',
+      license='GPL',
+      packages=['jmui'],
+      install_requires=['PyQt5!=5.15.0,!=5.15.1,!=5.15.2,!=6.0'],
```
Copied from `requirements/gui.txt`, but I'm not quite certain whether this is the right way to declare the dependency of PyQt5.

kristapsk replied:
See #913 for context, 5.15.0, 5.15.1, 5.15.2 and 6.0 had a bug for some users which is now fixed with newer versions.

wukong1971 commented on scripts/joinmarket-qt.py:
```diff
@@ -2305,6 +2305,15 @@ def onTabChange(i):
+openWalletDialog = QDialog()
+ui = Ui_openWalletDialog()
+ui.setupUi(openWalletDialog)
+openWalletDialog.show()
```
All the actual logic, like what happens when the user click a button, etc., will all be written at a separate place like in this `joinmarket-qt.py` file, instead of in the python UI file (like `open_wallet_dialog.py`).

AdamISZ commented:
Thanks for this. I agree with your general intention here, I think the initiative to use Qt designer will indeed be a helpful improvement (the app was initially developed by me with no experience in Qt work and it shows!), especially if we want the Qt app to develop further in future.

kristapsk commented:
Concept ACK on using Qt Designer.

wukong1971 commented:
This PR is ready for review now, I have also updated the description of this PR to share more details.
Thanks.

kristapsk commented on jmui/jmui/open_wallet_dialog.py:
```diff
@@ -0,0 +1,136 @@
+## WARNING! All changes made in this file will be lost when recompiling UI file!
+################################################################################
+
+from PySide2.QtCore import (QCoreApplication, QDate, QDateTime, QMetaObject,
+    QObject, QPoint, QRect, QSize, QTime, QUrl, Qt)
+from PySide2.QtGui import (QBrush, QColor, QConicalGradient, QCursor, QFont,
+    QFontDatabase, QIcon, QKeySequence, QLinearGradient, QPalette, QPainter,
+    QPixmap, QRadialGradient)
```
Unused imports should be removed.
```
$ ./test/lint/lint-python.sh 
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QDate' imported but unused
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QDateTime' imported but unused
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QObject' imported but unused
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QPoint' imported but unused
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QRect' imported but unused
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QTime' imported but unused
jmui/jmui/open_wallet_dialog.py:11:1: F401 'PySide2.QtCore.QUrl' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QBrush' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QColor' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QConicalGradient' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QCursor' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QFont' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QFontDatabase' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QIcon' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QKeySequence' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QLinearGradient' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QPalette' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QPainter' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QPixmap' imported but unused
jmui/jmui/open_wallet_dialog.py:13:1: F401 'PySide2.QtGui.QRadialGradient' imported but unused
```

wukong1971 replied:
This is generated code, these unused import will be re-introduced back during the next generation.
How about we exclude these generated code from the linter?

AdamISZ replied:
Indeed, that appears to be necessary with this workflow.

kristapsk commented:
Additionally I don't think `jmui` ir a good name here, `jmqtui` makes more sense to me, as cli is a UI too and there possiblly will be web based UI in future too, see https://github.com/JoinMarket-Org/jmcontrolcenter.

AdamISZ commented on jmui/setup.py:
```diff
@@ -0,0 +1,16 @@
+import os
+
+setup(name='joinmarketui',
+      version='0.9.0dev',
```
will need update. The root version string is in `JM_CORE_VERSION` in `jmbase/jmbase/support.py`. I guess we can't avoid multiple hard codings because they're package descriptions.

AdamISZ commented:
Agree with @kristapsk that we need `jmqtui` really, it's a little unclear otherwise.

A nit: handle the case of file not existing, something like is done here:

https://github.com/JoinMarket-Org/joinmarket-clientserver/blob/3b4a377c888194a6611dfa2b765985048a87dc76/scripts/joinmarket-qt.py#L1964-L1971

Apart from those small changes, this is a tACK from me. Appears to work as intended and is a good step. Thanks again.

wukong1971 commented:
PR updated:
1. Rebased to latest master.
2. Renamed `jmui` to `jmqtui`.
3. Updated `lint-python.sh` to ignore generated python files in `jmqtui/jmqtui` directory.
4. Updated the version in `setup.py` to '0.9.2dev'.
5. Added the exception handling of file not existing: show a QTMessageBox instead of crashing the app.

wukong1971 commented:
Hmm, so the unit test failed with the following error, I don't really understand what it means:
```
 >           raise InvalidSignature("forgot %s for %s" % (
                ', '.join(forgotten), self.commandName))
E           twisted.protocols.amp.InvalidSignature: forgot dust_threshold for b'JMInit'

jmvenv/lib/python3.8/site-packages/twisted/protocols/amp.py:1860: InvalidSignature
``

I don't think I have touched anything related to the signature, and I'm not sure how to fix this error here. Any help is highly appreciated, thanks in advance!

AdamISZ commented:
> I don't think I have touched anything related to the signature, and I'm not sure how to fix this error here. Any help is highly appreciated, thanks in advance!

Yes, this error was introduced into master by me, fixed today in ab9b1a83f585ab50247782febdf7d0bac20b5cfb.

Feel free to ignore (or rebase).

AdamISZ commented on jmqtui/setup.py:
```diff
@@ -0,0 +1,16 @@
+setup(name='joinmarketui',
+      version='0.9.2dev',
+      description='Joinmarket client library for Bitcoin coinjoins',
+      url='http://github.com/Joinmarket-Org/joinmarket-clientserver/jmfg',
```
"jmfg"?

wukong1971 replied:
Oops, nice catch, I'll fix it, thanks!

AdamISZ commented:
Apart from the url nit as per above, which is unimportant, I have one more nit:

If the initial wallet load is unsuccessful (and confirmed, handling invalid file paths is now correct), the UI reverts to its previous state and wallet loading will follow the old UI path, which is inconsistent (though, it functions fine according to my tests, i.e. you can still load another wallet later without error).

I think it's fine for now, but I presume you plan to make it consistent in future?

wukong1971 commented:
> I think it's fine for now, but I presume you plan to make it consistent in future?

Yes. This PR is only the beginning of improving the Qt UI of JoinMarket, I'm planning to spend much more effort in the future to continue improving the Qt UI. The end goal is that the user do not need a separate Electrum, Wasabi, or Spector to manage their coins, they can use JoinMarket as their only desktop Bitcoin wallet.

I did not change the whole workflow in this PR, because I'd like to keep the change as small as possible when we establish a new workflow of writing / updating the UI related code.

In this PR, we introduced a new `jmqtui` package, a new tool for designing the UI, as well as a new code generation flow. Once the architecture is in place, it'll be much easier to work with the UI in the following up PRs.

wukong1971 commented:
1. Rebased to latest master
2. Added a new commit to update the url in all `setup.py` files as the old links are no longer valid.

AdamISZ commented:
Thanks. tACK https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/932/commits/2ed3f21055a9124100a8f83439a734f19b679076 , merging.
